api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,254 @@
+    // the "response_trailers" field set.
+    HeaderResponse response_trailer_response = 6;
+
+    // The server can respond with this message instead of one of the above
+    // after a "request_headers" or "request_body" message. It causes an error
+    // to be immediately sent to the client. It is ignored after
+    // other message types.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. If multiple headers share the same key, they
+  // must be merged according to the HTTP spec. All header keys must be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool no_body = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a RequestHeaders or ResponseHeaders
+// message.
+message HeaderResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+}
+
+message BodyResponse {
+  // Add or replace HTTP headers.
```

Comment:
So, are we implicitly relying on the gRPC stream flow events (e.g. waiting for remote server response) to guide flow control (e.g. StopIteration, Continue) on the Envoy side? This makes sense, but I wonder if there is finer grained control over stop iteration, e.g. buffering, watermarking that we might need. In general I'd recommend that @alyssawilk takes a pass on this for watermark implications.

Reply:
I would assume we're doing the same thing we're doing for all async client operations here, which is relying on gRPC/H2 flow control for send and receive.  Overflow semantics seem documented on the request path, and I assume if any mutations on the gRPC resposne result in overflow we'd have the same close logic but it seems worth commenting.

Reply:
In general, I expect that, if the processing mode is "SYNC," then Envoy would send a message on the stream and suspend filter iteration until it gets back a response, so yes, that's the flow control mechanism. If the processing mode is ASYNC, however, then we want to avoid blowing up all the buffers in Envoy. I saw recently that each gRPC stream has a watermark on it, so perhaps that's a good way to go.

I will certainly need advice on the specific values to use, and those might best be done with a code review by @alyssawilk when we get to some implementation.